In [2]:
import numpy as np
import pandas as pd
import warnings
warnings.filterwarnings('ignore')

In [6]:
ukRA = pd.read_csv(r'd:\ELEC02\uk_road_accident.csv')
ukRA

Index Accident_Severity Accident Date   Latitude  \
0       200701BS64157           Serious      5/6/2019  51.506187   
1       200701BS65737           Serious      2/7/2019  51.495029   
2       200701BS66127           Serious    26-08-2019  51.517715   
3       200701BS66128           Serious    16-08-2019  51.495478   
4       200701BS66837            Slight      3/9/2019  51.488576   
...               ...               ...           ...        ...   
660674  201091NM01760            Slight    18-02-2022  57.374005   
660675  201091NM01881            Slight    21-02-2022  57.232273   
660676  201091NM01935            Slight    23-02-2022  57.585044   
660677  201091NM01964           Serious    23-02-2022  57.214898   
660678  201091NM02142           Serious    28-02-2022  57.575210   

                   Light_Conditions           District Area  Longitude  \
0             Darkness - lights lit  Kensington and Chelsea  -0.209082   
1                          Daylight  Kensington and Chelsea  -0.173647   
2       Darkness - lighting unknown  Kensington and Chelsea  -0.210215   
3                          Daylight  Kensington and Chelsea  -0.202731   
4             Darkness - lights lit  Kensington and Chelsea  -0.192487   
...                             ...                     ...        ...   
660674                     Daylight                Highland  -3.467828   
660675       Darkness - no lighting                Highland  -3.809281   
660676                     Daylight                Highland  -3.862727   
660677       Darkness - no lighting                Highland  -3.823997   
660678                     Daylight                Highland  -3.895673   

        Number_of_Casualties  Number_of_Vehicles Road_Surface_Conditions  \
0                          1                   2                     Dry   
1                          1                   2             Wet or damp   
2                          1                   3                     Dry   
3                          1                   4                     Dry   
4                          1                   2                     Dry   
...                      ...                 ...                     ...   
660674                     2                   1                     Dry   
660675                     1                   1            Frost or ice   
660676                     1                   3            Frost or ice   
660677                     1                   2             Wet or damp   
660678                     1                   1             Wet or damp   

                 Road_Type Urban_or_Rural_Area     Weather_Conditions  \
0       Single carriageway               Urban     Fine no high winds   
1       Single carriageway               Urban  Raining no high winds   
2                      NaN               Urban                    NaN   
3       Single carriageway               Urban     Fine no high winds   
4                      NaN               Urban                    NaN   
...                    ...                 ...                    ...   
660674  Single carriageway               Rural     Fine no high winds   
660675  Single carriageway               Rural     Fine no high winds   
660676  Single carriageway               Rural     Fine no high winds   
660677  Single carriageway               Rural     Fine no high winds   
660678    Dual carriageway               Rural  Snowing no high winds   

                                Vehicle_Type  
0                                        Car  
1                                        Car  
2                      Taxi/Private hire car  
3       Bus or coach (17 or more pass seats)  
4                              Other vehicle  
...                                      ...  
660674                                   Car  
660675                                   Car  
660676                                   Car  
660677                 Motorcycle over 500cc  
660

In [7]:
ukRA.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 660679 entries, 0 to 660678
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   Index                    660679 non-null  object 
 1   Accident_Severity        660679 non-null  object 
 2   Accident Date            660679 non-null  object 
 3   Latitude                 660654 non-null  float64
 4   Light_Conditions         660679 non-null  object 
 5   District Area            660679 non-null  object 
 6   Longitude                660653 non-null  float64
 7   Number_of_Casualties     660679 non-null  int64  
 8   Number_of_Vehicles       660679 non-null  int64  
 9   Road_Surface_Conditions  659953 non-null  object 
 10  Road_Type                656159 non-null  object 
 11  Urban_or_Rural_Area      660664 non-null  object 
 12  Weather_Conditions       646551 non-null  object 
 13  Vehicle_Type             660679 non-null  object 
dtypes: f

In [11]:
accident_severity_df = ukRA['Accident_Severity']
accident_severity_df

0         Serious
1         Serious
2         Serious
3         Serious
4          Slight
           ...   
660674     Slight
660675     Slight
660676     Slight
660677    Serious
660678    Serious
Name: Accident_Severity, Length: 660679, dtype: object

In [12]:
accident_date_df = ukRA['Accident Date']
accident_date_df

0           5/6/2019
1           2/7/2019
2         26-08-2019
3         16-08-2019
4           3/9/2019
             ...    
660674    18-02-2022
660675    21-02-2022
660676    23-02-2022
660677    23-02-2022
660678    28-02-2022
Name: Accident Date, Length: 660679, dtype: object

In [14]:
latitude_df = ukRA['Latitude']
latitude_df

0         51.506187
1         51.495029
2         51.517715
3         51.495478
4         51.488576
            ...    
660674    57.374005
660675    57.232273
660676    57.585044
660677    57.214898
660678    57.575210
Name: Latitude, Length: 660679, dtype: float64

In [15]:
light_conditions_df = ukRA['Light_Conditions']
light_conditions_df

0               Darkness - lights lit
1                            Daylight
2         Darkness - lighting unknown
3                            Daylight
4               Darkness - lights lit
                     ...             
660674                       Daylight
660675         Darkness - no lighting
660676                       Daylight
660677         Darkness - no lighting
660678                       Daylight
Name: Light_Conditions, Length: 660679, dtype: object

In [17]:
district_area_df = ukRA['District Area']
district_area_df

0         Kensington and Chelsea
1         Kensington and Chelsea
2         Kensington and Chelsea
3         Kensington and Chelsea
4         Kensington and Chelsea
                   ...          
660674                  Highland
660675                  Highland
660676                  Highland
660677                  Highland
660678                  Highland
Name: District Area, Length: 660679, dtype: object

In [18]:
longitude_df = ukRA['Longitude']
longitude_df

0        -0.209082
1        -0.173647
2        -0.210215
3        -0.202731
4        -0.192487
            ...   
660674   -3.467828
660675   -3.809281
660676   -3.862727
660677   -3.823997
660678   -3.895673
Name: Longitude, Length: 660679, dtype: float64

In [19]:
num_of_casualties = ukRA['Number_of_Casualties']
num_of_casualties

0         1
1         1
2         1
3         1
4         1
         ..
660674    2
660675    1
660676    1
660677    1
660678    1
Name: Number_of_Casualties, Length: 660679, dtype: int64

In [20]:
num_of_vehicles = ukRA['Number_of_Vehicles']
num_of_vehicles

0         2
1         2
2         3
3         4
4         2
         ..
660674    1
660675    1
660676    3
660677    2
660678    1
Name: Number_of_Vehicles, Length: 660679, dtype: int64

In [21]:
road_surface_conditions_df = ukRA['Road_Surface_Conditions']
road_surface_conditions_df

0                  Dry
1          Wet or damp
2                  Dry
3                  Dry
4                  Dry
              ...     
660674             Dry
660675    Frost or ice
660676    Frost or ice
660677     Wet or damp
660678     Wet or damp
Name: Road_Surface_Conditions, Length: 660679, dtype: object

In [22]:
road_type_df = ukRA['Road_Type']
road_type_df

0         Single carriageway
1         Single carriageway
2                        NaN
3         Single carriageway
4                        NaN
                 ...        
660674    Single carriageway
660675    Single carriageway
660676    Single carriageway
660677    Single carriageway
660678      Dual carriageway
Name: Road_Type, Length: 660679, dtype: object

In [23]:
area_df = ukRA['Urban_or_Rural_Area']
area_df

0         Urban
1         Urban
2         Urban
3         Urban
4         Urban
          ...  
660674    Rural
660675    Rural
660676    Rural
660677    Rural
660678    Rural
Name: Urban_or_Rural_Area, Length: 660679, dtype: object

In [24]:
weather_df = ukRA['Weather_Conditions']
weather_df

0            Fine no high winds
1         Raining no high winds
2                           NaN
3            Fine no high winds
4                           NaN
                  ...          
660674       Fine no high winds
660675       Fine no high winds
660676       Fine no high winds
660677       Fine no high winds
660678    Snowing no high winds
Name: Weather_Conditions, Length: 660679, dtype: object

In [25]:
vehicle_type_df = ukRA['Vehicle_Type']
vehicle_type_df

0                                          Car
1                                          Car
2                        Taxi/Private hire car
3         Bus or coach (17 or more pass seats)
4                                Other vehicle
                          ...                 
660674                                     Car
660675                                     Car
660676                                     Car
660677                   Motorcycle over 500cc
660678                                     Car
Name: Vehicle_Type, Length: 660679, dtype: object

<h1>QUESTIONS<h1>

<h2>1. How many accidents occured in total?<h2>

In [26]:
ukRA['Accident Date'].count()


np.int64(660679)

INSIGHTS: There are a total of 660679 accidents.

<h2>2.How many casualties were recorded in total?<h2>

In [31]:
ukRA['Number_of_Casualties'].sum()


np.int64(896568)

INSIGHTS: There are a total of 896568 casualties.

<h2>3. How many vehicles were involved in total?<h2>

In [32]:
ukRA['Number_of_Vehicles'].sum()

np.int64(1209872)

INSIGHTS: There are a total of 1209872 vehicles involved.

<h2>4. What is the average number of vehicles per accident?<h2>

In [38]:
ukRA['Number_of_Vehicles'].mean()

np.float64(1.8312554205597575)

INSIGHTS: Each accident involves about 1 to 2 vehicles.

<h2>5. What is the average number of casualties per accident?<h2>

In [37]:
ukRA['Number_of_Casualties'].mean()


np.float64(1.357040257068864)

INSIGHTS: Each accident involves 1 casualties.

<h2>6. Which accident severity is most common?<h2>

In [39]:
ukRA['Accident_Severity'].mode()

0    Slight
Name: Accident_Severity, dtype: object

INSIGHTS: The most common accident severity is slight.

<h2>7. Which road type has the most accidents?<h2>

In [40]:
ukRA['Road_Type'].mode()


0    Single carriageway
Name: Road_Type, dtype: object

INSIGHTS: The road type with the most accidents is Single carriageway.

<h2>8. Which weather condition is most common during accidents?<h2>

In [41]:
ukRA['Weather_Conditions'].mode()

0    Fine no high winds
Name: Weather_Conditions, dtype: object

INSIGHTS: Fine no high winds is the most commont weather condition during the accidents.

<h2>9. Which light condition has the most accidents?<h2>

In [43]:
ukRA['Light_Conditions'].mode()

0    Daylight
Name: Light_Conditions, dtype: object

INSIGHTS: Daylight has the most accidents.

<h2>10. Which road surface condition is most common?<h2>

In [44]:
ukRA['Road_Surface_Conditions'].mode()

0    Dry
Name: Road_Surface_Conditions, dtype: object

INSIGHTS: The most common road surface condition is Dry.

<h2>11.Average casualties per district area?<h2>

In [47]:
ukRA.groupby('District Area')['Number_of_Casualties'].mean()

District Area
Aberdeen City    1.139834
Aberdeenshire    1.310363
Adur             1.319871
Allerdale        1.474291
Alnwick          1.512931
                   ...   
Wychavon         1.427627
Wycombe          1.412377
Wyre             1.378029
Wyre Forest      1.382869
York             1.293094
Name: Number_of_Casualties, Length: 422, dtype: float64

INSIGHTS: Most districts average around 1 to 2 casualties per accident.

<h2>12. Total accidents per district area?<h2>

In [48]:
ukRA.groupby('District Area')['Accident Date'].count()

District Area
Aberdeen City    1323
Aberdeenshire    1930
Adur              619
Allerdale        1128
Alnwick           232
                 ... 
Wychavon         1361
Wycombe          1729
Wyre             1238
Wyre Forest       969
York             1897
Name: Accident Date, Length: 422, dtype: int64

INSIGHTS: District areas vary greatly in total accidents reported.

<h2>13. Average vehicles per district area?<h2>

In [49]:
ukRA.groupby('District Area')['Number_of_Vehicles'].mean()

District Area
Aberdeen City    1.637188
Aberdeenshire    1.553368
Adur             1.969305
Allerdale        1.782801
Alnwick          1.646552
                   ...   
Wychavon         1.828068
Wycombe          1.866397
Wyre             1.860258
Wyre Forest      1.789474
York             1.856089
Name: Number_of_Vehicles, Length: 422, dtype: float64

INSIGHTS: Most district areas involve around 1–2 vehicles per accident.

<h2>14. Casualties by accident severity?<h2>

In [50]:
ukRA.groupby('Accident_Severity')['Number_of_Casualties'].sum()

Accident_Severity
Fatal       16483
Serious    129439
Slight     750646
Name: Number_of_Casualties, dtype: int64

INSIGHTS: Slight accidents cause the highest number of casualties.

<h2>15. Vehicles by accident severity?<h2>

In [51]:
ukRA.groupby('Accident_Severity')['Number_of_Vehicles'].sum()

Accident_Severity
Fatal        15477
Serious     148057
Slight     1046338
Name: Number_of_Vehicles, dtype: int64

INSIGHTS: Slight accidents involve the highest number of vehicles.

<h2>16. How many casualties by road type?<h2>

In [52]:
ukRA.groupby('Road_Type')['Number_of_Casualties'].sum()

Road_Type
Dual carriageway      146877
One way street         16172
Roundabout             56085
Single carriageway    661768
Slip road              10024
Name: Number_of_Casualties, dtype: int64

INSIGHTS: Single carriageways have the most casualties.

<h2>17. Accidents by weather condition?<h2>

In [53]:
ukRA.groupby('Weather_Conditions')['Accident Date'].count()

Weather_Conditions
Fine + high winds          8554
Fine no high winds       520885
Fog or mist                3528
Other                     17150
Raining + high winds       9615
Raining no high winds     79696
Snowing + high winds        885
Snowing no high winds      6238
Name: Accident Date, dtype: int64

INSIGHTS: Fine weather with no high winds have the highest number of accidents.

<h2>18. Accidents by light condition?<h2>

In [55]:
ukRA.groupby('Light_Conditions')['Accident Date'].count()

Light_Conditions
Darkness - lighting unknown      6484
Darkness - lights lit          129335
Darkness - lights unlit          2543
Darkness - no lighting          37437
Daylight                       484880
Name: Accident Date, dtype: int64

INSIGHTS: Most accidents happen during daylight.

<h2>19. Which place has the most casualties, urban or rural?<h2>

In [56]:
ukRA.groupby('Urban_or_Rural_Area')['Number_of_Casualties'].sum()

Urban_or_Rural_Area
Rural          353515
Unallocated        13
Urban          543024
Name: Number_of_Casualties, dtype: int64

INSIGHTS: Urban areas have the highest number of casualties.

<h2>20. Average casualties by vehicle type?<h2>

In [57]:
ukRA.groupby('Vehicle_Type')['Number_of_Casualties'].mean()

Vehicle_Type
Agricultural vehicle                     1.342065
Bus or coach (17 or more pass seats)     1.349216
Car                                      1.358841
Data missing or out of range             1.500000
Goods 7.5 tonnes mgw and over            1.351881
Goods over 3.5t. and under 7.5t          1.362861
Minibus (8 - 16 passenger seats)         1.345648
Motorcycle 125cc and under               1.332635
Motorcycle 50cc and under                1.337235
Motorcycle over 125cc and up to 500cc    1.343521
Motorcycle over 500cc                    1.359434
Other vehicle                            1.340075
Pedal cycle                              1.370558
Ridden horse                             1.250000
Taxi/Private hire car                    1.368663
Van / Goods 3.5 tonnes mgw or under      1.354537
Name: Number_of_Casualties, dtype: float64

INSIGHTS: Cars have the highest average casualties per accident.